examples/Custom_Hooks/basics.py

```diff
@@ -0,0 +1,108 @@
+## Registering Custom Hooks
+
+# This example demonstrates the basic mechanics of the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility,
+# which lets you hook into any callable of your choice:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+
+from dataclasses import dataclass
+from time import perf_counter
+
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import RandomRecommender
+from baybe.searchspace import SearchSpace
+from baybe.utils.basic import register_hooks
+
+### Defining the Hooks
+
+# We start by defining a simple hook that lets us inspect the names of the parameters
+# involved in the recommendation process.
+# For this purpose, we match its signature to that of
+# {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`:
+#
+
+# ```{admonition} Signature components
+# :class: important
+# Note that, if provided, annotations must match **exactly** those of the target signature.
+# However, annotations are completely optional
+# — only the names/order of the signature parameters and their defaults matter.
+#
+# For the sake of demonstration, we only provide the annotation for the
+# relevant `searchspace` parameter of the hook in the example below.
+# ```
+
+
+def print_parameter_names_hook(
+    self,
+    batch_size,
+    searchspace: SearchSpace,
+    objective=None,
+    measurements=None,
+):
+    """Print the names of the parameters spanning the search space."""
+    print(f"Search space parameters: {[p.name for p in searchspace.parameters]}")
+
+
+# Additionally, we set up a class that provides a combination of hooks for measuring
+# the time needed to compute the recommendations:
+
+
+@dataclass
+class ElapsedTimePrinter:
+    """Helper class for measuring the time between two calls."""
+
+    last_call_time: float | None = None
+
+    def start(
+        instance,
+        self,
+        batch_size,
+        searchspace,
+        objective=None,
+        measurements=None,
+    ):
+        """Start the timer."""
+        instance.last_call_time = perf_counter()
+
+    def measure(
+        instance,
+        self,
+        batch_size,
+        searchspace,
+        objective=None,
+        measurements=None,
+    ):
+        """Measure the elapsed time."""
+        if instance.last_call_time is None:
+            raise RuntimeError("Must call `start` first!")
+        print(f"Elapsed time: {perf_counter() - instance.last_call_time}")
+
+
+### Monkeypatching
+
+# Next, we create our recommender and monkeypatch its `recommend` method:
+timer = ElapsedTimePrinter()
+recommender = RandomRecommender()
+RandomRecommender.recommend = register_hooks(
```

Comment:
is this order correct? even f it is, isnt it more natural (and safer order to suggest in a tutorial) to patch the class  before making an instance?

Reply:
I would say that's how you would typically do it in a real use case. Of course, you could also patch the class, but why would you want to do it? Typically, you want the effect only to take place on a specific object – so to me the current version is fine ... But let me know if you have a particular reason against it

Reply:
ah sorry, now I see what you are talking about!! wanted to write `recommender.recommend = register...`! Will fix stuff now

Reply:
I've changed things now. The current version is the best I could come up with. Unfortunately, if you really think it through what happens in the back (i.e. using bound methods vs functions, etc) the entire subject turns out to be not as straightforward as originally expected. Thus, the current version might appear unnecessarily complex to you at first glance, but perhaps you come to the same conclusion in the end. Let me know what you think ...

Reply:
approved